File: backend/models/Subject.js
import mongoose from "mongoose";
import Chapter from "./Chapter.js"; // Make sure path is correct

const subjectSchema = new mongoose.Schema({
  semesterId: { type: mongoose.Schema.Types.ObjectId, ref: "Semester", required: true },
  title: { type: String, required: true },
  courseCode: { type: String },
  instructor: { type: String },
  createdAt: { type: Date, default: Date.now }
});

// Return `id` instead of `_id` in JSON
subjectSchema.set('toJSON', {
  virtuals: true,
  versionKey: false,
  transform: (doc, ret) => { delete ret._id; }
});

/**
 * Pre-remove middleware to handle cascading deletes
 * Deleting a subject deletes all its chapters
 */
subjectSchema.pre("deleteOne", { document: true, query: false }, async function(next) {
  try {
    const chapters = await Chapter.find({ subjectId: this.id });
    // for (let chapter of chapters) {
    //   await chapter.deleteOne(); // triggers Chapter's pre-hook, which deletes Notes
    // }

    await Promise.all(chapters.map(chapter => chapter.deleteOne())); // 🔹 parallel delete

    next();
  } catch (err) {
    next(err);
  }
});

const Subject = mongoose.model("Subject", subjectSchema);

export default Subject;

File: frontend/lib/notesApiServices/Subjects/subject.dart
class Subject {
  final String? id;
  final String semesterId;
  final String title;
  final String courseCode;
  final String instructor;
  final DateTime createdAt;

  Subject({
    this.id,
    required this.semesterId,
    required this.title,
    this.courseCode = '',
    this.instructor = '',
    required this.createdAt,
  });

  factory Subject.fromJson(Map<String, dynamic> json) {
    return Subject(
      id: json['id'],
      semesterId: json['semesterId'],
      title: json['title'],
      courseCode: json['courseCode'] ?? '',
      instructor: json['instructor'] ?? '',
      createdAt: DateTime.parse(json['createdAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "id": id,
      "semesterId": semesterId,
      "title": title,
      "courseCode": courseCode,
      "instructor": instructor,
      "createdAt": createdAt.toIso8601String(),
    };
  }
}
